central/scannerdefinitions/handler/handler.go

```diff
@@ -535,96 +522,110 @@ func (h *httpHandler) post(w http.ResponseWriter, r *http.Request) {
-			log.Error(msg)
-			return errors.New(msg)
-		}
-		return nil
-	}
 
-	// Search for the Scanner V4 related file.
-	// It will have a specific prefix, but the full name is unknown.
-	var found bool
-	for _, zipF := range zipR.File {
-		if !strings.HasPrefix(zipF.Name, scannerV4DefsPrefix) {
-			continue
-		}
-		if found {
-			return errors.New("found multiple Scanner V4 versions")
-		}
-		if err := validate(zipF); err != nil {
-			return err
-		}
-		found = true
+		return mv.supportsRelease(mainVersionNum), nil
 	}
-	return nil
-}
 
-func (h *httpHandler) handleZipContentsFromVulnDump(ctx context.Context) error {
-	zipR, err := zip.OpenReader(h.uploadPath)
-	if err != nil {
-		return errors.Wrap(err, "couldn't open file as zip")
-	}
-	defer utils.IgnoreError(zipR.Close)
-	var count int
-	// It is expected a ZIP file be uploaded with both Scanner V2 and V4 vulnerability definitions.
-	// scanner-defs.zip contains data required by Scanner V2.
-	// scanner-v4-defs-*.zip contains data required by Scanner v4.
-	// In the future, we may decide to support other files (like we have in the past), which is why we
-	// loop through this ZIP of ZIPs.
+	// Map of offline definitions to add.
+	defs := make(map[string]*zip.File)
+
+	// Currently, we expect the offline ZIP to contain both Scanner V2 and V4 offline
+	// definitions. We want to keep the door open to support other file types, so
+	// keep the following loop sensible to structural changes in the zip.
 	for _, zipF := range zipR.File {
 		switch {
 		case zipF.Name == scannerV2DefsFile:
-			if err := h.handleScannerDefsFile(ctx, zipF, offlineScannerV2DefsBlobName); err != nil {
-				return errors.Wrap(err, "couldn't handle scanner-defs sub file")
+			if _, ok := defs[offlineScannerV2DefsBlobName]; ok {
+				return fmt.Errorf("v2 definitions: found duplicate definition: %s", zipF.Name)
+			}
+			log.Infof("v2 definitions: found supported bundle: %s", zipF.Name)
+			defs[offlineScannerV2DefsBlobName] = zipF
+		case features.ScannerV4.Enabled() && strings.HasPrefix(zipF.Name, scannerV4DefsPrefix):
+			ok, err := checkV4Manifest(zipF)
+			if err != nil {
+				// If we cannot check if the V4 definitions is supported, then we assume the
+				// bundle is not good, which is not entirely accurate but simplify error
+				// handling.
+				return fmt.Errorf("invalid v4 definition: %s: %w", zipF.Name, err)
+			}
+			if !ok {
+				// Ignore unsupported bundles.
+				log.Debugf("v4 definitions: ignoring unsupported bundle: %s", zipF.Name)
+				continue
 			}
-			log.Debugf("Successfully processed file: %s", zipF.Name)
-			count++
-		case strings.HasPrefix(zipF.Name, scannerV4DefsPrefix):
-			if err := h.handleScannerDefsFile(ctx, zipF, offlineScannerV4DefsBlobName); err != nil {
-				return errors.Wrap(err, "couldn't handle scanner-v4-defs sub file")
+			if prev, ok := defs[offlineScannerV4DefsBlobName]; ok {
+				return fmt.Errorf("v4 definitions: found more than one supported definitions: %s and %s",
```

Comment:
nit: definition or definition files

Reply:
Hum, I am considering a bundle within the archive a "definitions file", so maybe replace with "definitions file"?